http.go

```diff
@@ -346,93 +426,20 @@ func httpPutViolations(w http.ResponseWriter, r *http.Request) {
 		w.WriteHeader(http.StatusBadRequest)
 		return
 	}
-	httpPutViolationsInner(w, r, typestr, vs)
-}
-
-func httpPutViolationsInner(w http.ResponseWriter, r *http.Request, typestr string, vs []ViolationRequest) {
+	failedVs := []ViolationRequest{}
 	for _, v := range vs {
-		v.Fixup(typestr)
-		// Force type field to match value specified in request path
-		v.Type = typestr
-		err := v.Validate()
-		if err != nil {
-			log.Warnf(err.Error())
-			w.WriteHeader(http.StatusBadRequest)
-			return
-		}
-
-		if err = validateType(v.Type, v.Object); err != nil {
-			log.Warnf(err.Error())
-			w.WriteHeader(http.StatusBadRequest)
-			return
-		}
-
-		rep, err := repGet(typestr, v.Object)
-		if err == redis.Nil {
-			rep = Reputation{
-				Object:     v.Object,
-				Type:       typestr,
-				Reputation: 100,
-			}
-		} else if err != nil {
-			log.Warnf(err.Error())
-			w.WriteHeader(http.StatusInternalServerError)
-			return
-		}
-
-		err = rep.Validate()
-		if err != nil {
-			log.Warnf(err.Error())
-			w.WriteHeader(http.StatusInternalServerError)
-			return
-		}
-
-		// If recovery suppression was specified add the correct timestamp to the reputation
-		// entry. Is suppression is already indicated, only update it if it results in a new
-		// timestamp that is beyond what the existing value is.
-		if v.SuppressRecovery > 0 {
-			nd := time.Now().UTC().Add(time.Second *
-				time.Duration(v.SuppressRecovery))
-			if rep.DecayAfter.IsZero() || rep.DecayAfter.Before(nd) {
-				rep.DecayAfter = nd
-			}
+		if err := putViolation(typestr, v); err != nil {
+			failedVs = append(failedVs, v)
 		}
+	}
 
-		origRep := rep.Reputation
-		found, err := rep.applyViolation(v.Violation)
-		if err != nil {
-			if !found {
-				// Don't treat submitting an unknown violation as an error, instead
-				// just log it
-				log.WithFields(log.Fields{
-					"violation": v.Violation,
-					"object":    v.Object,
-					"type":      v.Type,
-				}).Warn("ignoring unknown violation")
-				continue
-			}
-			log.Warnf(err.Error())
-			w.WriteHeader(http.StatusInternalServerError)
-			return
-		}
-		err = rep.set()
-		if err != nil {
-			log.Warnf(err.Error())
-			w.WriteHeader(http.StatusInternalServerError)
-			return
-		}
-		exc := false
-		if rep.Type == "ip" {
-			exc = isException(rep.Object)
-		}
-		log.WithFields(log.Fields{
-			"violation":           v.Violation,
-			"object":              rep.Object,
-			"type":                rep.Type,
-			"reputation":          rep.Reputation,
-			"decay_after":         rep.DecayAfter,
-			"original_reputation": origRep,
-			"exception":           exc,
-		}).Info("violation applied")
+	respByt, err := json.Marshal(failedVs)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
 	}
+
+	w.Write(respByt)
+	w.WriteHeader(http.StatusOK)
```

Comment:
I think we should either:
1. Return a `400` if there are any violation failures
2. Not include a response body if there are no failures.

or potentially both.

It seems weird that you have to unmarshal and check the contents of the body to see if there are any failures.

Reply:
now doing both ✅